src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -163,5 +171,63 @@ public static void AppendStatusTrailers(HttpResponse response, Status status)
             response.AppendTrailer(GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
             response.AppendTrailer(GrpcProtocolConstants.MessageTrailer, status.Detail);
         }
+
+        public static AuthContext CreateAuthContext(X509Certificate2 clientCertificate)
```

Comment:
@jiangtaoli2016   this is an implementation of AuthContext that  should be compatible with C core's implementation - can you take a look?

Reply:
@jiangtaoli2016 What kind of validation does gRPC C core do with the client certificate?

Reply:
The certificate validation logic happens in `tsi_create_ssl_client_handshaker_factory_with_options()` and `tsi_create_ssl_server_handshaker_factory_with_options()` in `ssl_transport_security.cc` for client and server.

Reply:
IIUC `CreateAuthContext()` implements the C core logic of parsing X.509 certificate into `tsi_peer_property` and then `grpc_auth_context`. If that's the case, I think we further need to take into account peer properties such as ALPN and session reuse if they are supported.

Reply:
So part of my concerns here are the internal guidelines in our SDL for client certification validation, but also, honestly, I'm rather confused by the enumeration of SANs. Sure, it's subject alternative name, but really, it's DNS entries. It just struck me as odd. Is the intent here that the majority of folks just use their HTTPS cert?

Reply:
To clarify, is the auth piece really adjacent? I noticed in the spec it's more a case of "Whatever is presented we'll put in the context" to allow for other formats, so if I read it right, the auth, certainly with regards to certificates is not a grpc concern, but a server concern.